app/uk/gov/hmrc/cdsreimbursementclaimfrontend/controllers/upload/FileUpload.scala

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2021 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.cdsreimbursementclaimfrontend.controllers.upload
+
+import cats.implicits.{catsSyntaxEq, catsSyntaxOptionId}
+import com.google.inject.{Inject, Singleton}
+import play.api.Configuration
+import play.api.mvc.Call
+import uk.gov.hmrc.cdsreimbursementclaimfrontend.config._
+import uk.gov.hmrc.cdsreimbursementclaimfrontend.models.answers.SupportingEvidencesAnswer
+import uk.gov.hmrc.cdsreimbursementclaimfrontend.models.upscan.{ScheduledDocument, SupportingEvidence, UploadReference, UpscanUpload}
+import uk.gov.hmrc.cdsreimbursementclaimfrontend.controllers.supportingevidence.{routes => evidenceRoutes}
+import uk.gov.hmrc.cdsreimbursementclaimfrontend.models.JourneyStatus.FillingOutClaim
+import uk.gov.hmrc.cdsreimbursementclaimfrontend.models.upscan.UpscanCallBack.UpscanSuccess
+
+trait FileUpload[A] {
+
+  val key: String
+
+  val maxUploads: Int
+
+  def hasReachedUploadThreshold(maybeAnswer: Option[A]): Boolean
+
+  def hasReference(maybeAnswer: Option[A], uploadReference: UploadReference): Boolean
+
+  def add(
+    upload: UpscanUpload,
+    callback: UpscanSuccess,
+    maybeAnswer: Option[A],
+    claim: FillingOutClaim
+  ): FillingOutClaim
+
+  def reviewPage: Call
+
+  def uploadErrorPage: Call
+
+  def handleUploadCallback(uploadReference: UploadReference): Call
+
+  def scanSuccessPage(uploadReference: UploadReference): Call
+
+  def scanErrorPage: Call
+}
+
+@Singleton
+class FileUploadServices @Inject() (config: Configuration) {
```

Comment:
Let's rename this to `FileUploadInstance` - you're almost moving towards a type-class design again here.